File: frontend/src/api/petBattle.ts
import api from './client';

// ========================================
// 宠物对战系统 API
// ========================================

export interface PetBattleInfo {
  pet_id: number;
  name: string;
  species: string;
  level: number;
  evolution_stage: number;
  hp: number;
  max_hp: number;
  combo: number;
  ultimate_charges: number;
}

export interface QuestionData {
  word_id: number;
  word: string;
  question_text: string;
  options: string[];
}

export interface RoundResult {
  round_number: number;
  question: QuestionData;
  player1_answer: string | null;
  player1_correct: boolean;
  player1_time_ms: number | null;
  player1_damage: number;
  player1_used_ultimate: boolean;
  player1_hp_after: number;
  player2_answer: string | null;
  player2_correct: boolean;
  player2_time_ms: number | null;
  player2_damage: number;
  player2_used_ultimate: boolean;
  player2_hp_after: number;
}

export interface Battle {
  id: number;
  status: string;
  mode: string;
  current_round: number;
  max_rounds: number;
  player1_id: number;
  player1_username: string;
  player1_pet: PetBattleInfo;
  player1_total_correct: number;
  player1_total_damage: number;
  player2_id: number;
  player2_username: string;
  player2_pet: PetBattleInfo;
  player2_total_correct: number;
  player2_total_damage: number;
  winner_id: number | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  expires_at: string | null;
}

export interface BattleListItem {
  id: number;
  opponent_username: string;
  opponent_pet_name: string;
  status: string;
  mode: string;
  result: string | null;
  created_at: string;
}

export interface BattleStats {
  total_battles: number;
  wins: number;
  losses: number;
  draws: number;
  win_rate: number;
  current_win_streak: number;
  max_win_streak: number;
  total_damage_dealt: number;
  total_damage_taken: number;
  avg_damage_per_battle: number;
  accuracy: number;
  ultimates_used: number;
  ultimates_landed: number;
  perfect_wins: number;
  comeback_wins: number;
  rating: number;
  peak_rating: number;
}

export interface CreateBattleRequest {
  opponent_id: number;
  wordbook_id?: number;
  mode?: string;
  max_rounds?: number;
}

// ========== HTTP API ==========

export const createBattle = async (data: CreateBattleRequest): Promise<Battle> => {
  return api.post('/student/battle/create', data);
};

export const acceptBattle = async (battleId: number): Promise<Battle> => {
  return api.post(`/student/battle/${battleId}/accept`);
};

export const cancelBattle = async (battleId: number): Promise<{ message: string }> => {
  return api.post(`/student/battle/${battleId}/cancel`);
};

export const getBattle = async (battleId: number): Promise<Battle> => {
  return api.get(`/student/battle/${battleId}`);
};

export const getMyBattles = async (status?: string, limit = 20): Promise<BattleListItem[]> => {
  return api.get('/student/battles/my', { params: { status, limit } });
};

export const getPendingInvites = async (): Promise<Battle[]> => {
  return api.get('/student/battles/invites');
};

export const getBattleStats = async (): Promise<BattleStats> => {
  return api.get('/student/battles/stats');
};

// ========== WebSocket ==========

export interface WSMessage {
  type: string;
  [key: string]: any;
}

export class BattleWebSocket {
  private ws: WebSocket | null = null;
  private listeners: Map<string, Set<(data: any) => void>> = new Map();
  private reconnectTimeout: ReturnType<typeof setTimeout> | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private battleId: number;
  private token: string;

  constructor(battleId: number, token: string) {
    this.battleId = battleId;
    this.token = token;
  }

  connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      const wsUrl = `${import.meta.env.VITE_WS_URL || 'ws://localhost:8000'}/api/v1/student/battle/ws/${this.battleId}?token=${this.token}`;

      this.ws = new WebSocket(wsUrl);

      this.ws.onopen = () => {
        console.log('Battle WebSocket connected');
        this.reconnectAttempts = 0;
        resolve();
      };

      this.ws.onmessage = (event) => {
        try {
          const data = JSON.parse(event.data);
          this.emit(data.type, data);
        } catch (error) {
          console.error('Failed to parse WS message:', error);
        }
      };

      this.ws.onerror = (error) => {
        console.error('Battle WebSocket error:', error);
        reject(error);
      };

      this.ws.onclose = () => {
        console.log('Battle WebSocket closed');
        this.emit('close', {});
        this.attemptReconnect();
      };
    });
  }

  on(event: string, callback: (data: any) => void) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set());
    }
    this.listeners.get(event)!.add(callback);
  }

  off(event: string, callback: (data: any) => void) {
    const listeners = this.listeners.get(event);
    if (listeners) {
      listeners.delete(callback);
    }
  }

  private emit(event: string, data: any) {
    const listeners = this.listeners.get(event);
    if (listeners) {
      listeners.forEach((callback) => callback(data));
    }
  }

  send(data: any) {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(data));
    } else {
      console.error('WebSocket not connected');
    }
  }

  private attemptReconnect() {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      console.error('Max reconnect attempts reached');
      return;
    }

    this.reconnectAttempts++;
    const delay = Math.min(1000 * Math.pow(2, this.reconnectAttempts), 10000);

    console.log(`Attempting reconnect in ${delay}ms...`);

    this.reconnectTimeout = setTimeout(() => {
      this.connect().catch((error) => {
        console.error('Reconnect failed:', error);
      });
    }, delay);
  }

  close() {
    if (this.reconnectTimeout) {
      clearTimeout(this.reconnectTimeout);
    }
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
}

// ========== 答题WebSocket(独立通道) ==========

export class AnswerWebSocket {
  private ws: WebSocket | null = null;
  private battleId: number;
  private token: string;

  constructor(battleId: number, token: string) {
    this.battleId = battleId;
    this.token = token;
  }

  connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      const wsUrl = `${import.meta.env.VITE_WS_URL || 'ws://localhost:8000'}/api/v1/student/battle/ws/${this.battleId}/answer?token=${this.token}`;

      this.ws = new WebSocket(wsUrl);

      this.ws.onopen = () => {
        console.log('Answer WebSocket connected');
        resolve();
      };

      this.ws.onerror = (error) => {
        console.error('Answer WebSocket error:', error);
        reject(error);
      };
    });
  }

  submitAnswer(roundNumber: number, answer: string, timeMs: number, useUltimate = false) {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(
        JSON.stringify({
          round_number: roundNumber,
          answer,
          time_ms: timeMs,
          use_ultimate: useUltimate,
        })
      );
    }
  }

  close() {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
}

File: backend/app/schemas/pet_battle.py
"""宠物对战系统 - Pydantic Schema"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


# ========== 请求模型 ==========

class BattleCreateRequest(BaseModel):
    """创建对战请求"""
    opponent_id: int = Field(..., description="对手用户ID")
    wordbook_id: Optional[int] = Field(None, description="单词本ID，为空则随机")
    mode: str = Field("casual", description="模式: casual/ranked")
    max_rounds: int = Field(10, description="最大回合数")


class BattleAnswerRequest(BaseModel):
    """提交答案请求"""
    battle_id: int
    round_number: int
    answer: str = Field(..., description="A/B/C/D")
    use_ultimate: bool = Field(False, description="是否使用必杀技")


# ========== 响应模型 ==========

class PetBattleInfo(BaseModel):
    """宠物战斗信息"""
    pet_id: int
    name: str
    species: str
    level: int
    evolution_stage: int
    hp: int
    max_hp: int
    combo: int
    ultimate_charges: int

    class Config:
        from_attributes = True


class QuestionData(BaseModel):
    """题目数据"""
    word_id: int
    word: str
    question_text: str
    options: List[str]  # ["A. 快乐的", "B. 悲伤的", "C. 生气的", "D. 害怕的"]


class RoundResult(BaseModel):
    """回合结果"""
    round_number: int
    question: QuestionData

    # 玩家1
    player1_answer: Optional[str]
    player1_correct: bool
    player1_time_ms: Optional[int]
    player1_damage: int
    player1_used_ultimate: bool
    player1_hp_after: int

    # 玩家2
    player2_answer: Optional[str]
    player2_correct: bool
    player2_time_ms: Optional[int]
    player2_damage: int
    player2_used_ultimate: bool
    player2_hp_after: int


class BattleResponse(BaseModel):
    """对战详情响应"""
    id: int
    status: str
    mode: str
    current_round: int
    max_rounds: int

    # 玩家1
    player1_id: int
    player1_username: str
    player1_pet: PetBattleInfo
    player1_total_correct: int
    player1_total_damage: int

    # 玩家2
    player2_id: int
    player2_username: str
    player2_pet: PetBattleInfo
    player2_total_correct: int
    player2_total_damage: int

    # 胜负
    winner_id: Optional[int]

    # 时间
    created_at: datetime
    started_at: Optional[datetime]
    finished_at: Optional[datetime]
    expires_at: Optional[datetime]

    class Config:
        from_attributes = True


class BattleListItem(BaseModel):
    """对战列表项"""
    id: int
    opponent_username: str
    opponent_pet_name: str
    status: str
    mode: str
    result: Optional[str]  # win/lose/draw
    created_at: datetime

    class Config:
        from_attributes = True


class BattleStatsResponse(BaseModel):
    """对战统计响应"""
    total_battles: int
    wins: int
    losses: int
    draws: int
    win_rate: float

    current_win_streak: int
    max_win_streak: int

    total_damage_dealt: int
    total_damage_taken: int
    avg_damage_per_battle: float

    accuracy: float  # 正确率

    ultimates_used: int
    ultimates_landed: int

    perfect_wins: int
    comeback_wins: int

    rating: int
    peak_rating: int

    class Config:
        from_attributes = True


# ========== WebSocket 消息 ==========

class WSBattleStart(BaseModel):
    """战斗开始消息"""
    type: str = "battle_start"
    battle: BattleResponse


class WSNewRound(BaseModel):
    """新回合消息"""
    type: str = "new_round"
    round_number: int
    question: QuestionData
    time_limit: int  # 秒


class WSAnswerReceived(BaseModel):
    """收到答案消息"""
    type: str = "answer_received"
    player_id: int
    round_number: int


class WSRoundResult(BaseModel):
    """回合结果消息"""
    type: str = "round_result"
    result: RoundResult


class WSBattleEnd(BaseModel):
    """战斗结束消息"""
    type: str = "battle_end"
    winner_id: Optional[int]
    winner_name: Optional[str]

    # 奖励
    food_earned: int
    xp_earned: int
    rating_change: Optional[int]

    # 统计
    player1_final_stats: dict
    player2_final_stats: dict


class WSError(BaseModel):
    """错误消息"""
    type: str = "error"
    message: str
    code: Optional[str] = None
